src/lib/api.ts

```diff
@@ -27,9 +35,152 @@ export async function createGroup(groupFormValues: GroupFormValues) {
+Hence Double insertions for the pending transaction.
+
+Same condition will be there for update. One person is updating the transaction and other is inserting in DB.
+
+Locking ensures consistency and correctness.
+
+Retrying is there to retry acquiring locks after some time if it was acquired by someone else at that instant.
+Discussion: https://github.com/spliit-app/spliit/pull/114/files#r1559974206
+*/
+
+export async function acquireLockRecurringTransaction(groupId:string, expenseId:string, lockId: string): Promise<RecurringTransactions[]> {
+  const prisma = await getPrisma()
+  let retryCnt = 5
+  let getRecTxn:RecurringTransactions[] = []
+  while(--retryCnt) {
+    const existingLock = await prisma.recurringTransactions.findMany({
+      where: {
+        groupId,
+        expenseId
+      }
+    })
+    if (!existingLock.length) {
+      const query = `
+        INSERT INTO "RecurringTransactions"
+          ("groupId", "expenseId", "createNextAt", "lockId")
+          VALUES
+          ('${groupId}', '${expenseId}', 0, '${lockId}')
+          ON CONFLICT("groupId", "expenseId")
+          DO NOTHING;
+      `
+      await prisma.$queryRawUnsafe(query)
+    }
+    await prisma.recurringTransactions.updateMany({
+      where: {
+          groupId,
+          expenseId,
+          lockId: null
+      },
+      data: {
+        lockId
+      }
+    })
+    getRecTxn = await prisma.recurringTransactions.findMany({
+      where: {
+        expenseId: expenseId,
+        groupId,
+        lockId
+      }
+    })
+    const receivedLockId = getRecTxn?.[0]?.lockId
+    if (receivedLockId === lockId) break;
+    await sleep(500);
+  }
+  return getRecTxn
+}
+
+export async function releaseLockRecurringTransaction(groupId:string, expenseId:string, lockId: string): Promise<void>{
+  const prisma = await getPrisma()
+  await prisma.recurringTransactions.updateMany({
+    where: {
+      groupId,
+      expenseId: expenseId,
+      lockId,
+    },
+    data: {
+      lockId: null
+    }
+  })
+}
+
+export async function createOrUpdateRecurringTransaction(
+  expenseFormValues: ExpenseFormValues,
+  groupId: string,
+  expenseId: string,
+  expenseRef: string|null = null
+): Promise<RecurringTransactions | undefined> {
+  let expenseDate = expenseFormValues.expenseDate
+  const prisma = await getPrisma()
+  if (!+expenseFormValues.recurringDays) {
+    await prisma.recurringTransactions.deleteMany({
+      where: {
+        groupId,
+        expenseId: expenseRef || expenseId
+      }
+    })
+    return Promise.resolve(undefined)
+  }
+  const lockId:string = randomId();
+  const receivedLockId = (await acquireLockRecurringTransaction(groupId, expenseRef || expenseId, lockId))?.[0]?.lockId
+  let recTxn;
+  if (!receivedLockId) return recTxn
+  const epochTime = Math.floor((new Date(expenseDate)).getTime() / 1000)
+  const nextAt: number = epochTime + (+expenseFormValues.recurringDays * 86400)
```

Comment:
This creates the transaction exactly 24 hours after the previous transaction. It might be ever-so-slightly better to use `Math.floor(epochTime / 86400) + (1 + expenseFormValues.recurringDays) * 86400` to get midnight of the day of the next occurrence.

Reply:
Sure.
Better way to do this is by using crons [as mentioned in this comment](https://github.com/spliit-app/spliit/pull/114#discussion_r1615333622). I can add this here or in V2 PR where I'll be changing everything from Days to crons?